### PR TITLE
Introduce SWIO

### PR DESCRIPTION
**Pull request by rmontuoro**
This PR introduces the SWPC's new I/O component ([SWIO](https://github.com/NOAA-SWPC/SWIO)) and updates to IPE, WAM, and the SWPC mediator. These updates are documented in the respective PRs: [#27](https://github.com/NOAA-SWPC/IPE/pull/27) (IPE), [#18](https://github.com/NOAA-SWPC/WAM/pull/18) (WAM), and [#1](https://github.com/NOAA-SWPC/NEMS/pull/1).

### CODE REVIEW

File: scripts/compsets/config/wcoss.phase2.config [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (akubaryk) [inline scripts/compsets/config/wcoss.phase2.config:1]: We've dropped Phase 2 support; this and the symlinked phase 1 can be deleted.

File: scripts/compsets/exglobal/exglobal_fcst_nems.sh [hunk 2]
```diff
@@ -1422,11 +1438,335 @@ runSeq::
     @$coupling_interval_fast_sec
       ATM
     @
+    ATM -> AIO
```
Comment (akubaryk) [inline scripts/compsets/exglobal/exglobal_fcst_nems.sh:1441]: need to add AIO. might be better to have the IO by default in an outer loop?

File: scripts/compsets/exglobal/exglobal_fcst_nems.sh [hunk 3]
```diff
@@ -1422,11 +1438,335 @@ runSeq::
...
     IPM
+    IO
+    IPM -> IO
   @
 ::
 EOF
 
+cat << EOF > $DATA/med.rc
+#############################################
+####  SWIO Mediator Configuration File  #####
```
Comment (akubaryk) [inline scripts/compsets/exglobal/exglobal_fcst_nems.sh:1451]: SWPC Mediator

File: scripts/compsets/exglobal/exglobal_fcst_nems.sh [hunk 4]
```diff
@@ -1422,11 +1438,335 @@ runSeq::
...
+#
+# If true, this flag transforms the vertical coordinate (v),
+# assumed in km, to heights (h) relative to the average
+# Earth's radius (R=6,371.2 km), accordina to the formula:
+#   h = 1 + v/R
+#
+output_grid_level_relative: false
+EOF
 
 else
```
Comment (akubaryk) [inline scripts/compsets/exglobal/exglobal_fcst_nems.sh:1772]: Think you mentioned we can remove the cp of wam3dgridnew*.nc files. Is ipe3dgrid2.nc still necessary?

Comment (akubaryk) [inline scripts/compsets/exglobal/exglobal_fcst_nems.sh:1772]: We should link the new output files from RUNDIR to ROTDIR; I can help with this if needed.

File: scripts/compsets/config/wcoss.phase2.config (hunk 1 above)
Comment (rmontuoro) [reply scripts/compsets/config/wcoss.phase2.config:1]: Agreed. Please proceed with the change.

File: scripts/compsets/exglobal/exglobal_fcst_nems.sh (hunk 2 above)
Comment (rmontuoro) [reply scripts/compsets/exglobal/exglobal_fcst_nems.sh]: I've added AIO to run at the same frequency as IO. This run sequence is just intended to demonstrate how to use SWIO. We may refine the run sequence at a later time, when the requirements will be further defined.

File: scripts/compsets/exglobal/exglobal_fcst_nems.sh (hunk 3 above)
Comment (rmontuoro) [reply scripts/compsets/exglobal/exglobal_fcst_nems.sh]: Fixed.

File: scripts/compsets/exglobal/exglobal_fcst_nems.sh (hunk 4 above)
Comment (rmontuoro) [reply scripts/compsets/exglobal/exglobal_fcst_nems.sh:1773]: From the application's standpoint this file is no longer required.

Comment (rmontuoro) [reply scripts/compsets/exglobal/exglobal_fcst_nems.sh:1773]: Agreed. Please make all necessary changes.

**Comment by twfang** [comment]
I have plotted out some initial comparisons. Thes plots come from Raffaele's run with mesh pushing above 800km.  
Op_TEC
TEC_diff

**Comment by twfang** [comment]
Several other comparisons as well. Here, old refers to George's method (i.e. old postprocessing) and new refers to the new IO. 
OLD:
old_1
NEW:
new_1
OLD2:
old_2
NEW2:
new_2

They are very similar, however, I notice a slight shift between the new and old.

**Comment by twfang** [comment]
It turns out the slight shift I showed in the meeting comes from plotting. Here are the updated figures. 
 
OLD:
old_fixed

NEW:
new_fixed

**Comment by twfang** [comment]
I have also corrected the TEC difference plots. 
Op_TEC

With interpolation, now I am seeing about 5 TECU difference between these methods. 
TEC_diff

**Comment by ZhuxiaoLi66** [comment]
The one hour WAM io output (2013031600-01) with 3min cedence by the io branch (SWIO) has been compared with the WAM original output in the same run except by the develop branch in GSMWAM-IPE.
The Neutral Temperature & horizontal wind at about 250km and 400km (lev135 and lev144) has been compared as following.

**Comment by ZhuxiaoLi66** [comment]
T_U_V_400km_wam_origin_output_2013031601
T_U_V_400km_wam_io_output_2013031601

**Comment by ZhuxiaoLi66** [comment]
T_U_V_250km_wam_origin_output_2013031601_1
T_U_V_250km_wam_io_output_2013031601_1

**Comment by twfang** [comment]
@ZhuxiaoLi66 , for these plots did you process the sig files from the same run where IO outputs the gsm files?

**Comment by ZhuxiaoLi66** [comment]
In brief, the neutral temperature of 'io' output (gsm. 20130316*nc) seems about 100 degree colder than the regular output of WAM, I also noticed that the minimum Temp is 0. The winds looks closer while need further investgation.
Will look the neutral density later this afternoon.
The other minor things about the Netcdf format of the output (gsm.*nc) are that there is no time dimenson, significant coordination for each variable.

**Comment by ZhuxiaoLi66** [comment]
@twfang they are the NetCDF format output, it is the same with sig files. The run is based on develop branch, should be bitwise same, will do the comparison later .

**Comment by twfang** [comment]
@ZhuxiaoLi66 , see if you can also use the NetCDF files from the same run with the IO for the comparison. Just in case there are other changes from this branch.

**Comment by ZhuxiaoLi66** [comment]
@twfang you are correct. The output of that two are quite not the bitwise same. The following T_U_V plot are using the orgin output from the same io branch run.  the results seems much closer.
T_U_V_400km_wam_origin_output_by_io_branch
T_U_V_250km_wam_origin_output_by_io_branch

**Comment by twfang** [comment]
So, the outputs using the original method and IO in this branch are consistent, right?

Could you please check out the develop branch with the same commits in the IO branch to see what causes the differences? There were some updates on scripts in develop branch that Adam made in early April.

**Comment by ZhuxiaoLi66** [comment]
I think we can say that the IO WAM output are consistent with the regular WAM output. 
The following is the O number density comparison. 
O_numden_250km_WAM_io_output
O_numden_250km_WAM_origin_output_by_io_branch

**Comment by ZhuxiaoLi66** [comment]
I will compare several WAM branch later. 
@rmontuoro I didn't find the vairalbe of 'mms' called 'mean mass' (as Tzu-Wei said) in the io output or the 'pool export fields' file.  the other very important validation variable is 'den', total number density.

**Comment by twfang** [comment]
Zhuxiao, these parameters need to be further calculated since they are not standard WAM outputs. Once IO branch is committed to the develop branch, we can fine-tune the output frequency or add some of these parameters.

**Comment by akubaryk** [comment]
They are calculated, that's how they're a part of the wam_fields.nc output -- we just need to know how to hook those arrays into the WAM export fields (and possibly change the frequency of calculation depending on some details).

**Comment by ZhuxiaoLi66** [comment]
got it, thanks. 
I compared the orgin WAM output from the io branch and the develop branch in GSMWAM-IPE, the results are very close, showed as following, so there should be no concern about the merge at this perspective. 
T_U_V_250km_wam_origin_output_by_io_branch
T_U_V_250km_WAM_origin_output_develop

**Comment by ZhuxiaoLi66** [comment]
Actually, bitwise same.

**Comment by twfang** [comment]
Thanks Zhuxiao!

**Comment by twfang** [comment]
Zhuxiao, I have changed you to one of the reviewers. If you think WAM results look good, please go ahead and approve the PR.

**Comment by twfang** [comment]
I have also looked at the hourly electric fields from IPE results. They generally look good and results are what we expected. These plots are in geographic longitude and latitude. Values plotted here are from 350 km. 

Zonal drift (Eastward positive)
eastward_dft

Meridional drift (Northward positive)
northward_dft

Vertical drift (Upward positive)
up_dft

Vertical drift at geographic equator
dft_equ

**Comment by ZhuxiaoLi66** [comment]
Thanks, Tzu-Wei. the only concern about the WAM io output is that the minimum value of neutral Temp. is 0. the values in origin output is about 800 degree at 250km. @rmontuoro would like to know that have you set the values of some specific points into zero? thanks.

**Comment by twfang** [comment]
@rmontuoro , please let us know where the 0 in WAM temp might be coming from. 

I run the same version at WCOSS, and all the runs went well.

**Comment by ZhuxiaoLi66** [comment]
I checked the io NetCDF files,  it looks like Rafaele set the variables values over south pole and north pole (2*90=180 values) equal zeros. maybe we can discuss and revise it in later version. 
Another minor thing is that the io outputs don't have the ones at the very beginning time and ending time. e.g.   gsm.20130316_000000.nc and gsm.20130316_010000.nc for 2013031600-01 run. 
maybe also need to retune in next version. 
In brief, the main results are good for WAM, I am going to approve the merge.

**Pull request merged**